mono/metadata/w32handle.c

```diff
@@ -21,26 +21,23 @@
 #include "utils/mono-proclib.h"
 #include "utils/mono-threads.h"
 #include "utils/mono-time.h"
+#include "utils/mono-mmap.h"
 
 #undef DEBUG_REFS
 
-#define SLOT_MAX		(2048)
-
-/* must be a power of 2 */
-#define HANDLE_PER_SLOT	(4096)
+typedef struct _MonoW32HandleSlot MonoW32HandleSlot;
+struct _MonoW32HandleSlot {
+	MonoW32HandleSlot *next;
+/* Make MonoW32HandleSlot 8kB */
+#define HANDLES_PER_SLOT ((8192 - /* MonoW32HandleSlot.next */ sizeof(MonoW32HandleSlot*)) / sizeof(MonoW32Handle))
```

Comment:
This page-table-size trick is wrong and will waste a lot of memory on systems will 64k pages like aarch64.

Reply:
You pick the page sizes in this case, not the processor architecture.

You pick the number of levels and the sizes. And you can have different sizes at different levels (just like processors do). A page size like 64k is nice when you need many total bits overall -- more bits per level equals fewer levels, but we would need fewer bits than typical paging systems. We also can have more levels than a typical page table also, because they have fewer levels to shorten the walks, since they are walking them frequently. We would walk infrequently, so more levels are ok, which cuts memory use.